### PR TITLE
JSONRPC Async Model

### PR DESCRIPTION
**Pull request by taooceros**
Question:
1. why not use json serialize method in ToString method in JSONRPCModel.
2. whether we want an async version of action?
3. async version for context menu?

### CODE REVIEW

**Comment by taooceros** [comment]
Got a question.

```c#
if (result.StartsWith("DEBUG:"))
{
    MessageBox.Show(new Form {TopMost = true}, result.Substring(6));
    return string.Empty;
}
```

Does anyone actually use this to debug? This prevents making it based on Stream, because Stream are impossible to peek.

**Comment by jjw24** [comment]
Does this just show a message box when result starts with debug?

**Comment by taooceros** [comment]
> Does this just show a message box when result starts with debug?

Yes

**Comment by taooceros** [comment]
It will prevent from assuming the returning message is based on JsonRPCQueryResponseModel so that we can't deserialize it based on stream.

**Comment by jjw24** [comment]
Yeah let's get rid of it. Surely there is a better implementation for debugging

**Comment by jjw24** [comment]
Ready for review?

**Comment by taooceros** [comment]
Ah yes!

**Comment by jjw24** [comment]
The to dos in the PR description are no longer relevant right?

**Comment by taooceros** [comment]
> The to dos in the PR description are no longer relevant right?

Yeah they are questions, not quite relate to the change in this branch.

**Comment by jjw24** [comment]
Have we done any testing with this?

**Comment by jjw24** [comment]
> > The to dos in the PR description are no longer relevant right?
> 
> Yeah they are questions, not quite relate to the change in this branch.

Hmmm i havent played enough with non-c# plugins to be able to answer those. General rule of thumb, if it improves Flow and you have tested it to make sure, then lets PR it. Leaving those questions here I will likely forget, unless you will remember :)

**Comment by taooceros** [comment]
> Have we done any testing with this?

I check out that python plugin will work if python works. If It doesn't work, the error reporting part will be triggered (I think the EndofStream works).

File: Flow.Launcher.Core/Plugin/JsonRPCPlugin.cs [hunk 1]
```diff
@@ -27,23 +27,10 @@ internal abstract class JsonRPCPlugin : IPlugin, IContextMenu
         /// </summary>
         public abstract string SupportedLanguage { get; set; }
 
-        protected abstract string ExecuteQuery(Query query);
+        protected abstract Task<Stream> ExecuteQueryAsync(Query query, CancellationToken token);
```
Comment (jjw24) [inline Flow.Launcher.Core/Plugin/JsonRPCPlugin.cs:30]: Is this change backwards compatible?

File: Flow.Launcher.Core/Plugin/JsonRPCPlugin.cs [hunk 2]
```diff
@@ -130,70 +135,123 @@ private void ExecuteFlowLauncherAPI(string method, object[] parameters)
...
-                Log.Exception($"|JsonRPCPlugin.Execute|Exception for filename <{startInfo.FileName}> with argument <{startInfo.Arguments}>", e);
-                return string.Empty;
+                Log.Exception(
+                    $"|JsonRPCPlugin.ExecuteAsync|Exception for filename <{startInfo.FileName}> with argument <{startInfo.Arguments}>",
+                    e);
+                return Stream.Null;
+            }
+        }
+
+        public async Task<List<Result>> QueryAsync(Query query, CancellationToken token)
```
Comment (jjw24) [inline Flow.Launcher.Core/Plugin/JsonRPCPlugin.cs:237]: the change from Query to QueryAsync, will this be backwards compatible?

**Comment by taooceros** [comment]
> Hmmm i havent played enough with non-c# plugins to be able to answer those. General rule of thumb, if it improves Flow and you have tested it to make sure, then lets PR it. Leaving those questions here I will likely forget, unless you will remember :)

Well actually these questions are also relate to c# plugins. What do you think about them?

File: Flow.Launcher.Core/Plugin/JsonRPCPlugin.cs (hunk 1 above)
Comment (taooceros) [reply Flow.Launcher.Core/Plugin/JsonRPCPlugin.cs:30]: The JSONRPCPlugin is in Flow, and what it is doing is that overriding the Query (or QueryAsync) to call the executable (or python) and communicate via console stream. So change here won't affect the compatibility of plugins.

**Comment by taooceros** [comment]
The change in this branch are majorly using the IAsyncPlugin to allow early stop (not completely), and read Stream instead of reading string to reduce the cost of creating string based on Standard output stream.

**Comment by jjw24** [comment]
> > Hmmm i havent played enough with non-c# plugins to be able to answer those. General rule of thumb, if it improves Flow and you have tested it to make sure, then lets PR it. Leaving those questions here I will likely forget, unless you will remember :)
> 
> Well actually these questions are also relate to c# plugins. What do you think about them?

oh yeah you are right. I think why not, we can do it for QueryAsync implementation and test it, not that it's a necessity right now though, but implementing it may allow for future usage and expansion.

**Comment by jjw24** [comment]
Let's create issues for the questions and  can discuss or implement it.

**Review by jjw24 — APPROVED** [review]

**Pull request merged**